logbook-spring-webflux/src/main/java/org/zalando/logbook/spring/webflux/LogbookWebFilter.java

```diff
@@ -30,24 +30,27 @@
 
         return Mono
                 .just(exchange)
-                .doOnNext((e) -> triggerStageChange(stage, serverRequest, serverResponse))
+                .doOnNext((e) -> triggerStageChange(stage, serverRequest, serverResponse, false))
                 .map(e -> exchange
                         .mutate()
-                        .request(new BufferingServerHttpRequest(e.getRequest(), serverRequest, bytes -> triggerStageChange(stage, serverRequest, serverResponse)))
-                        .response(new BufferingServerHttpResponse(e.getResponse(), serverResponse, bytes -> triggerStageChange(stage, serverRequest, serverResponse)))
+                        .request(new BufferingServerHttpRequest(e.getRequest(), serverRequest, bytes -> triggerStageChange(stage, serverRequest, serverResponse, false)))
+                        .response(new BufferingServerHttpResponse(e.getResponse(), serverResponse, bytes -> triggerStageChange(stage, serverRequest, serverResponse, false)))
                         .build()
                 )
                 .flatMap(chain::filter)
-                .doOnSuccess(it -> triggerStageChange(stage, serverRequest, serverResponse)) // ensure response has been written
+                .doOnSuccess(it -> triggerStageChange(stage, serverRequest, serverResponse, true)) // ensure response has been written
                 .then();
     }
 
-    private void triggerStageChange(AtomicReference<Object> stage, ServerRequest request, ServerResponse response) {
-        stage.updateAndGet(throwingUnaryOperator(s -> {
-            if (s instanceof Logbook) return ((Logbook) s).process(request);
-            if (s instanceof Logbook.RequestWritingStage) return ((Logbook.RequestWritingStage) s).write().process(response);
-            if (s instanceof Logbook.ResponseWritingStage) ((Logbook.ResponseWritingStage) s).write();
-            return null;
-        }));
+    private void triggerStageChange(AtomicReference<Object> stage, ServerRequest request, ServerResponse response, Boolean forceFinish) {
```

Comment:
I can't remember reviewing the original version of this method. There has to be a cleaner way to do this...

Reply:
Do you mean this stage transition or do-while? If it's regarding the cycle, I've removed it.